rfpremier/ah-django
# #164857235 Track and view comment edit history

### What does this PR do?

 - [x] Allow models to track create, update and delete history of a comment
 - [x] Update comments Model
 - [x] Add functionality to view previous of a comment
 
 
### Tasks to be complete
 - [x]  A user should be able to view the edit history of a comment

### How can this be manually tested?
- Clone the repo and checkout to the  branch `ft/164857235-comment-history`
- Install, create and activate a virtual environment and ensure you have running Postgres database
- Run `pip install -r requirements.txt`
- Setup environment variables including URL to your Postgres database 
- Run `python manage.py makemigrations`
- Run `python manage.py migrate`
- If you are using an existing database, run `python manage.py populate_history --auto`
- Run the dev server `python manage.py runserver`

#### On Postman test the following endpoints:
- First signup if you don't have an account `[POST] http://127.0.0.1:8000/api/users/`
Required fields
```
{
            "user": {
                "email": "testuser@gmail.com",
                "username": "testuser",
                "password": "testuserpass123"
            }
}
```
Response
```{
    "user": {
        "email": "testuser@gmail.com",
        "username": "testuser",
        "token": "token"
    }
}
```
**Use the token provided for authentication to access some of the endpoints below**

- Post an article `[POST] http://127.0.0.1:8000/api/articles/`
- Fill in the authentication token
Required fields
```
{
            "article": {
                "title": "This is the article title",
                "description": "This is the article description",
                "body": "This is the article body",
                "image_url": "https://imageurl.com"
            }

        }
``` 
Response
```
{
    "id": 1,
    "title": "This is the article title",
    "description": "This is the article description",
    "body": "This is the article body",
    "image_url": "https://imageurl.com",
    "author": {
            "username": "testuser",
            "bio": "",
            "image": ""
        },
    "created_at": "2019-04-04T12:20:34.114095Z",
    "updated_at": "2019-04-04T12:20:34.114468Z",
    "slug": "this-is-the-article-title"
}
```
- POST comment `[POST] http://127.0.0.1:8000/api/articles/this-is-the-article-title/comments`
Request Body
```
 {
	"comment" : {"body" : "comment here"}
}
```
Response
```
{
    "id": 1,
    "created_at": "2019-04-17T22:37:58.997585Z",
    "updated_at": "2019-04-17T22:37:58.997628Z",
    "body": "Comment here",
    "user": {
        "username": "testuser",
        "bio": "",
        "image": ""
    },
    "highlighted_text": "",
    "article": "this-is-the-article-title"
}
```

- Edit a comment`[PUT]  http://127.0.0.1:8000/api/articles/this-is-the-article-title/comments/1`
Request Body
```
 {
	"comment" : {"body" : "edited comment here" }
}
```
Response
```
 {
    "id": 1,
    "created_at": "2019-04-17T22:37:58.997585Z",
    "updated_at": "2019-04-17T22:37:58.997628Z",
    "body": "edited comment here",
    "user": {
        "username": "testuser",
        "bio": "",
        "image": ""
    },
    "highlighted_text": "",
    "article": "this-is-the-article-title"
}
```

- To view the edit history of the comment, head to `[GET]  http://127.0.0.1:8000/api/comments/1/history`

Response 
```
{
    "comment_id": 1,
    "comment_history": [
        {
            "updated_at": "2019-04-17T22:37:58.997628Z",
            "history_id": 1,
            "body": "Edited comment here"
        }
    ]
}
```

 ### PT stories relevant to this PR?
[#164857235](https://www.pivotaltracker.com/story/show/164857235)
